gluten-core/src/main/scala/org/apache/spark/util/TaskResources.scala

```diff
@@ -60,20 +51,97 @@ object TaskResources extends TaskListener with Logging {
+                  return
+                }
+                val tid = Thread.currentThread().getId
+                val registry = RESOURCE_REGISTRIES.remove(context)
+                // When task failed and contains sub-thread computation,
+                // we should release sub-thread's resources first.
+                registry.asScala.filter(_._1 != tid).foreach(_._2.releaseAll())
+                registry.asScala.filter(_._1 == tid).foreach(_._2.releaseAll())
+              }
+              // TODO:
+              // The general duty of printing error message should not reside in memory module
+              error match {
+                case e: TaskKilledException if e.reason == "another attempt succeeded" =>
+                case _ => logError(s"Task ${context.taskAttemptId()} failed by error: ", error)
+              }
+            }
+          })
+        tc.addTaskCompletionListener(new TaskCompletionListener {
+          override def onTaskCompletion(context: TaskContext): Unit = {
+            RESOURCE_REGISTRIES.synchronized {
+              // All queries were fallback
+              if (!RESOURCE_REGISTRIES.containsKey(tc)) {
+                return
+              }
+              val registry = RESOURCE_REGISTRIES.remove(context)
+              assert(
+                registry.size() == 1,
+                "Current Spark TaskContext contains extra registry at task completion!")
+              registry.get(Thread.currentThread().getId).releaseAll()
+              context
+                .taskMetrics()
+                .incPeakExecutionMemory(
+                  registry.get(Thread.currentThread().getId).getSharedMetrics().peak())
+            }
+          }
+        })
+        return RESOURCE_REGISTRIES.get(tc).get(currentTId)
+      } else {
+        // Found a sub-thread computation under same Spark task's scope, e.g. PythonRunner.
+        val registryMap = RESOURCE_REGISTRIES.get(tc)
+        if (!registryMap.containsKey(currentTId)) {
+          logInfo(
+            s"Current thread is a sub-thread of Spark Task thread, " +
+              s"need create a new resource registry.")
+          val glutenTaskContext = new JniTaskContext
+          val registry = new TaskResourceRegistry(currentTId, currentTName)
+          registry.needExplicitRelease = true
+          registry.addResource(currentTId.toString, glutenTaskContext)
+          registryMap.put(currentTId, registry)
+        }
+        registryMap.get(Thread.currentThread().getId)
+      }
+    }
+  }
+
+  def releaseCurrentResources(): Unit = {
+    RESOURCE_REGISTRIES.synchronized {
+      val currentTId = Thread.currentThread().getId
+      val registryMap = RESOURCE_REGISTRIES.get(getLocalTaskContext())
+      if (registryMap.containsKey(currentTId) && registryMap.get(currentTId).needExplicitRelease) {
+        val registry = registryMap.remove(currentTId)
+        logDebug(s"Release all resources in ${registry.name}")
+        registry.releaseAll()
```

Comment:
So by the condition `needExplicitRelease`, `releaseCurrentResources` only releases the resources from off-task thread, does it?

Reply:
Yes.